node/impl/full/multisig.go

```diff
@@ -0,0 +1,271 @@
+		}
+		src = defaddr
+	}
+
+	if gp == types.EmptyInt {
+		gp = types.NewInt(1)
+	}
+
+	// Set up constructor parameters for multisig
+	msigParams := &samsig.ConstructorParams{
+		Signers:               addrs,
+		NumApprovalsThreshold: req,
+	}
+
+	enc, actErr := actors.SerializeParams(msigParams)
+	if actErr != nil {
+		return api.UndefAddressPair, actErr
+	}
+
+	// new actors are created by invoking 'exec' on the init actor with the constructor params
+	execParams := &init_.ExecParams{
+		CodeCID:           builtin.MultisigActorCodeID,
+		ConstructorParams: enc,
+	}
+
+	enc, actErr = actors.SerializeParams(execParams)
+	if actErr != nil {
+		return api.UndefAddressPair, actErr
+	}
+
+	// now we create the message to send this with
+	msg := types.Message{
+		To:       builtin.InitActorAddr,
+		From:     src,
+		Method:   builtin.MethodsInit.Exec,
+		Params:   enc,
+		GasPrice: gp,
+		GasLimit: 1000000,
+		Value:    val,
+	}
+
+	// send the message out to the network
+	smsg, err := a.MpoolAPI.MpoolPushMessage(ctx, &msg)
+	if err != nil {
+		return api.UndefAddressPair, err
+	}
+
+	// wait for it to get mined into a block
+	wait, err := a.StateAPI.StateWaitMsg(ctx, smsg.Cid())
+	if err != nil {
+		return api.UndefAddressPair, err
+	}
+
+	// check it executed successfully
+	if wait.Receipt.ExitCode != 0 {
+		fmt.Println("actor creation failed!")
+		return api.UndefAddressPair, err
+	}
+
+	// get address of newly created miner
+
+	var execreturn init_.ExecReturn
+	if err := execreturn.UnmarshalCBOR(bytes.NewReader(wait.Receipt.Return)); err != nil {
+		return api.UndefAddressPair, err
+	}
+	fmt.Println("Created new multisig: ", execreturn.IDAddress, execreturn.RobustAddress)
+
+	return api.AddressPair{
+		IDAddress:     execreturn.IDAddress,
+		RobustAddress: execreturn.RobustAddress,
+	}, nil
+}
+
+func (a *MsigAPI) MsigPropose(ctx context.Context, msig address.Address, to address.Address, amt types.BigInt, src address.Address, method uint64, params string) (api.MsigProposeOutput, error) {
+
+	if msig == address.Undef {
+		return api.UndefMsigProposalOutput, xerrors.Errorf("must provide a multisig address for proposal")
+	}
+
+	if to == address.Undef {
+		return api.UndefMsigProposalOutput, xerrors.Errorf("must provide a target address for proposal")
+	}
+
+	if !types.BigPos(amt) {
+		return api.UndefMsigProposalOutput, xerrors.Errorf("must provide a positive amount for proposed send")
+	}
+
+	if src == address.Undef {
+		defaddr, err := a.WalletAPI.WalletDefaultAddress(ctx)
+		if err != nil {
+			return api.UndefMsigProposalOutput, err
+		}
+		src = defaddr
+	}
+
+	paramBytes, err := hex.DecodeString(params)
+	if err != nil {
+		return api.UndefMsigProposalOutput, err
+	}
+
+	enc, actErr := actors.SerializeParams(&samsig.ProposeParams{
+		To:     to,
+		Value:  amt,
+		Method: abi.MethodNum(method),
+		Params: paramBytes,
+	})
+	if actErr != nil {
+		return api.UndefMsigProposalOutput, actErr
+	}
+
+	msg := &types.Message{
```

Comment:
There might be some DRYing opportunities here.

Reply:
Hmm how would you recommend I go about that? The only DRY I really practiced in this PR was moving the cli logic into the RPC calls.

Reply:
I don't have any immediate suggestions for DRYing. In cases like these, I just look at the code and try to determine if there are any patterns, like two or more calls making, say, three or more calls that are exactly the same or only vary in one parameter or so. It's a little tricky. There's a fine line between DRYing that makes sense and DRYing that can potentially hamper maintenance due to creating functions that make little logical sense. If you don't see anything, don't sweat it. :)

Reply:
I just spent some time comparing the three functions. The three functions seem too different to merge logic. If anything I was hoping to find some similarities between Propose and Approve, but most of the logic is input validation and serialization which differs since they take different parameters :/ I'm thinking if I add a `Cancel` function, it should be easy to factor out some common code with `Propose` since they should take the same params. I'll start to tackle that one now.

Reply:
Fair enough. Nothing immediately jumped out at me after looking a little more closely but I thought there might be a chance. That said, yes, I think `Approve`/`Cancel` should have more opportunities since they're essentially twins, at least in terms of what goes out on the wire.